File: monitoring-system/server/app/services/visitor_service.py
"""
Visitor log service — CRUD, assignment and replies.
Kept independent from the case-submission service by design.
"""
from datetime import datetime, timezone
from ..extensions import db
from ..models.visitor_log import VisitorLog, VisitorReply
from ..models.user import User


def create_visitor(data: dict) -> tuple[dict, int]:
    visitor = VisitorLog(
        name=data["name"],
        company=data.get("company") or None,
        visit_date=data["visit_date"],
        time_in=data["time_in"],
        reason_for_visit=data["reason_for_visit"],
        description=data.get("description") or None,
    )
    db.session.add(visitor)
    db.session.commit()
    return {"message": "Thank you. Your visit has been logged.", "visitor": visitor.to_dict()}, 201


def get_visitors(page: int = 1, per_page: int = 20, status: str = None) -> tuple[dict, int]:
    if page < 1 or per_page < 1 or per_page > 100:
        return {"error": "page must be >= 1 and per_page must be between 1 and 100."}, 400
    query = VisitorLog.query
    if status:
        query = query.filter_by(status=status)
    paginated = query.order_by(VisitorLog.created_at.desc()).paginate(
        page=page, per_page=per_page, error_out=False
    )
    return {
        "visitors": [v.to_dict() for v in paginated.items],
        "total":    paginated.total,
        "pages":    paginated.pages,
        "page":     page,
        "per_page": per_page,
    }, 200


def get_visitor(visitor_id: int) -> tuple[dict, int]:
    visitor = VisitorLog.query.get_or_404(visitor_id)
    return {"visitor": visitor.to_dict(include_replies=True)}, 200


def update_visitor_status(visitor_id: int, status: str) -> tuple[dict, int]:
    visitor = VisitorLog.query.get_or_404(visitor_id)
    visitor.status = status
    db.session.commit()
    return {"message": "Status updated.", "visitor": visitor.to_dict()}, 200


def assign_visitor(visitor_id: int, assigned_to: int, actor: User) -> tuple[dict, int]:
    visitor = VisitorLog.query.get_or_404(visitor_id)
    staff = User.query.get_or_404(assigned_to)
    if staff.role not in ("secretary", "admin") or not staff.is_active:
        return {"error": "Target user is not an active staff member."}, 400

    visitor.assigned_to = assigned_to
    visitor.assigned_by = actor.id
    visitor.assigned_at = datetime.now(timezone.utc)
    if visitor.status == "pending":
        visitor.status = "assigned"
    db.session.commit()
    return {"message": "Visitor assigned.", "visitor": visitor.to_dict()}, 200


def delete_visitor(visitor_id: int) -> tuple[dict, int]:
    visitor = VisitorLog.query.get_or_404(visitor_id)
    db.session.delete(visitor)
    db.session.commit()
    return {"message": "Visitor log deleted."}, 200


def add_reply(visitor_id: int, responder_id: int, message: str) -> tuple[dict, int]:
    visitor = VisitorLog.query.get_or_404(visitor_id)
    reply = VisitorReply(visitor_id=visitor.id, responder_id=responder_id, message=message)
    db.session.add(reply)
    db.session.commit()
    return {"message": "Reply added.", "reply": reply.to_dict()}, 201

File: monitoring-system/server/src/services/visitorService.js
import { prisma } from '../db.js';
import { HttpError } from '../utils/httpError.js';
import { visitorDict, visitorReplyDict } from '../utils/serialize.js';
import { parseHHMM, parseDateOnly } from '../utils/dates.js';

const REPLIES_INCLUDE = {
  replies: { include: { responder: true }, orderBy: { created_at: 'asc' } },
  assigned_user: true,
};

async function findOr404(id, include) {
  const v = await prisma.visitor_logs.findUnique({ where: { id }, include });
  if (!v) throw new HttpError(404, 'Resource not found.');
  return v;
}

export async function createVisitor(data) {
  const visitor = await prisma.visitor_logs.create({
    data: {
      name: data.name,
      company: data.company || null,
      visit_date: parseDateOnly(data.visit_date),
      time_in: parseHHMM(data.time_in),
      reason_for_visit: data.reason_for_visit,
      description: data.description || null,
    },
  });
  return {
    status: 201,
    body: { message: 'Thank you. Your visit has been logged.', visitor: visitorDict(visitor) },
  };
}

export async function getVisitors(page = 1, perPage = 20, status) {
  if (page < 1 || perPage < 1 || perPage > 100) {
    throw new HttpError(400, 'page must be >= 1 and per_page must be between 1 and 100.');
  }
  const where = status ? { status } : {};
  const [total, items] = await prisma.$transaction([
    prisma.visitor_logs.count({ where }),
    prisma.visitor_logs.findMany({
      where,
      include: { assigned_user: true },
      orderBy: { created_at: 'desc' },
      skip: (page - 1) * perPage,
      take: perPage,
    }),
  ]);
  return {
    visitors: items.map((v) => visitorDict(v)),
    total,
    pages: total === 0 ? 0 : Math.ceil(total / perPage),
    page,
    per_page: perPage,
  };
}

export async function getVisitor(id) {
  const v = await findOr404(id, REPLIES_INCLUDE);
  return { visitor: visitorDict(v, { includeReplies: true }) };
}

export async function updateVisitorStatus(id, status) {
  await findOr404(id);
  const v = await prisma.visitor_logs.update({
    where: { id }, data: { status }, include: { assigned_user: true },
  });
  return { message: 'Status updated.', visitor: visitorDict(v) };
}

export async function assignVisitor(id, assignedTo, actor) {
  const visitor = await findOr404(id);
  const staff = await prisma.users.findUnique({ where: { id: assignedTo } });
  if (!staff) throw new HttpError(404, 'Resource not found.');
  if (!['secretary', 'admin'].includes(staff.role) || !staff.is_active) {
    throw new HttpError(400, 'Target user is not an active staff member.');
  }
  const v = await prisma.visitor_logs.update({
    where: { id },
    data: {
      assigned_to: assignedTo,
      assigned_by: actor.id,
      assigned_at: new Date(),
      status: visitor.status === 'pending' ? 'assigned' : visitor.status,
    },
    include: { assigned_user: true },
  });
  return { message: 'Visitor assigned.', visitor: visitorDict(v) };
}

export async function deleteVisitor(id) {
  await findOr404(id);
  await prisma.$transaction(async (tx) => {
    await tx.visitor_replies.deleteMany({ where: { visitor_id: id } });
    await tx.visitor_logs.delete({ where: { id } });
  });
  return { message: 'Visitor log deleted.' };
}

export async function addReply(id, responderId, message) {
  await findOr404(id);
  const reply = await prisma.visitor_replies.create({
    data: { visitor_id: id, responder_id: responderId, message },
    include: { responder: true },
  });
  return { status: 201, body: { message: 'Reply added.', reply: visitorReplyDict(reply) } };
}
